File: src/diagnostics.rs
#[cfg(unix)]
use std::fs;
use std::path::Path;

use rusqlite::Connection;
use serde::Serialize;

use crate::discovery;

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Diagnostic {
    pub level: &'static str,
    pub check: &'static str,
    pub message: String,
}

pub fn run(config_path: &Path, database_path: &Path) -> Vec<Diagnostic> {
    let mut checks = Vec::new();
    if config_path.exists() {
        checks.push(Diagnostic {
            level: "ok",
            check: "configuration",
            message: format!("Config found: {}", config_path.display()),
        });
        if let Some(check) = permission_check(config_path, "config permissions") {
            checks.push(check);
        }
        match discovery::firecrawl_status(config_path) {
            Ok((false, _)) => checks.push(Diagnostic {
                level: "ok",
                check: "Firecrawl",
                message: "Optional Firecrawl discovery is disabled.".into(),
            }),
            Ok((true, true)) => checks.push(Diagnostic {
                level: "ok",
                check: "Firecrawl",
                message: "Enabled and FIRECRAWL_API_KEY is present in the environment.".into(),
            }),
            Ok((true, false)) => checks.push(Diagnostic {
                level: "error",
                check: "Firecrawl",
                message: "Enabled but FIRECRAWL_API_KEY is not set in the environment.".into(),
            }),
            Err(error) => checks.push(Diagnostic {
                level: "error",
                check: "Firecrawl",
                message: format!("Invalid Firecrawl configuration: {error:#}"),
            }),
        }
    } else {
        checks.push(Diagnostic {
            level: "warn",
            check: "configuration",
            message: format!(
                "Config not found: {}. The tracker can still use --database or the default path.",
                config_path.display()
            ),
        });
    }

    if !database_path.exists() {
        checks.push(Diagnostic {
            level: "warn",
            check: "database",
            message: format!("Database does not exist yet: {}", database_path.display()),
        });
        return checks;
    }

    match Connection::open(database_path) {
        Ok(connection) => {
            match connection.query_row("PRAGMA user_version", [], |row| row.get::<_, i64>(0)) {
                Ok(3) => checks.push(Diagnostic {
                    level: "ok",
                    check: "database schema",
                    message: "Schema 3 is current and Python-compatible.".into(),
                }),
                Ok(version) if version > 3 => checks.push(Diagnostic {
                    level: "error",
                    check: "database schema",
                    message: format!("Schema {version} is newer than supported schema 3."),
                }),
                Ok(version) => checks.push(Diagnostic {
                    level: "warn",
                    check: "database schema",
                    message: format!("Schema {version} will migrate to schema 3 when opened."),
                }),
                Err(error) => checks.push(Diagnostic {
                    level: "error",
                    check: "database schema",
                    message: error.to_string(),
                }),
            }
            match connection.query_row("PRAGMA quick_check", [], |row| row.get::<_, String>(0)) {
                Ok(value) if value == "ok" => checks.push(Diagnostic {
                    level: "ok",
                    check: "database integrity",
                    message: "SQLite quick_check: ok".into(),
                }),
                Ok(value) => checks.push(Diagnostic {
                    level: "error",
                    check: "database integrity",
                    message: value,
                }),
                Err(error) => checks.push(Diagnostic {
                    level: "error",
                    check: "database integrity",
                    message: error.to_string(),
                }),
            }
        }
        Err(error) => checks.push(Diagnostic {
            level: "error",
            check: "database",
            message: format!("Cannot inspect {}: {error}", database_path.display()),
        }),
    }
    if let Some(check) = permission_check(database_path, "database permissions") {
        checks.push(check);
    }
    checks
}

#[cfg(unix)]
fn permission_check(path: &Path, check: &'static str) -> Option<Diagnostic> {
    use std::os::unix::fs::PermissionsExt;
    let mode = fs::metadata(path).ok()?.permissions().mode() & 0o777;
    if mode & 0o077 != 0 {
        Some(Diagnostic {
            level: "warn",
            check,
            message: format!(
                "{} is mode {mode:04o}; consider restricting it to 0600.",
                path.display()
            ),
        })
    } else {
        Some(Diagnostic {
            level: "ok",
            check,
            message: format!("{} is restricted to the current user.", path.display()),
        })
    }
}

#[cfg(not(unix))]
fn permission_check(_path: &Path, _check: &'static str) -> Option<Diagnostic> {
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn missing_paths_produce_actionable_warnings() {
        let checks = run(
            Path::new("/definitely/missing/openjobscout/config.toml"),
            Path::new("/definitely/missing/openjobscout/jobs.sqlite3"),
        );
        assert!(checks.iter().any(|check| check.check == "configuration"));
        assert!(checks.iter().any(|check| check.check == "database"));
    }
}

File: src/open_job_scout/diagnostics.py
from __future__ import annotations

import importlib.util
import os
import sqlite3
import stat
from dataclasses import asdict, dataclass
from pathlib import Path

from .config import DEFAULT_CONFIG, expand_path, load_config
from .database import SCHEMA_VERSION
from .firecrawl import settings_from_config


@dataclass(frozen=True, slots=True)
class Diagnostic:
    level: str
    check: str
    message: str

    def as_dict(self) -> dict[str, str]:
        return asdict(self)


def _permission_check(path: Path, label: str) -> Diagnostic | None:
    if os.name == "nt" or not path.exists():
        return None
    mode = stat.S_IMODE(path.stat().st_mode)
    if mode & 0o077:
        return Diagnostic(
            "warn",
            f"{label} permissions",
            f"{path} is mode {mode:04o}; consider restricting it to 0600.",
        )
    return Diagnostic("ok", f"{label} permissions", f"{path} is restricted to the current user.")


def _writable_parent(path: Path) -> Path | None:
    candidate = path if path.exists() and path.is_dir() else path.parent
    while not candidate.exists() and candidate != candidate.parent:
        candidate = candidate.parent
    return candidate if candidate.exists() and candidate.is_dir() else None


def run_diagnostics(config_path: Path | None = None) -> list[Diagnostic]:
    selected = (config_path or DEFAULT_CONFIG).expanduser().resolve()
    checks: list[Diagnostic] = []
    if not selected.exists():
        return [
            Diagnostic(
                "error",
                "configuration",
                f"Config not found: {selected}. Run `jobscout init` first.",
            )
        ]

    try:
        config = load_config(selected)
    except (OSError, ValueError) as exc:
        return [Diagnostic("error", "configuration", str(exc))]

    checks.append(Diagnostic("ok", "configuration", f"Valid config: {selected}"))
    permission = _permission_check(selected, "config")
    if permission:
        checks.append(permission)

    sites = [str(site).lower() for site in config["search"]["sites"]]
    if "indeed" in sites:
        checks.append(
            Diagnostic(
                "error",
                "sources",
                "Indeed is configured but currently disabled because the upstream adapter "
                "does not verify TLS certificates.",
            )
        )
    else:
        checks.append(Diagnostic("ok", "sources", f"Configured sources: {', '.join(sites)}"))

    firecrawl = settings_from_config(config)
    if not firecrawl.enabled:
        checks.append(
            Diagnostic("ok", "Firecrawl", "Optional Firecrawl discovery is disabled.")
        )
    elif os.environ.get("FIRECRAWL_API_KEY", "").strip():
        checks.append(
            Diagnostic(
                "ok",
                "Firecrawl",
                "Enabled and FIRECRAWL_API_KEY is present in the environment.",
            )
        )
    else:
        checks.append(
            Diagnostic(
                "error",
                "Firecrawl",
                "Enabled but FIRECRAWL_API_KEY is not set in the environment.",
            )
        )

    database = expand_path(config["storage"]["database"])
    if database.exists():
        try:
            with sqlite3.connect(database, timeout=2.0) as connection:
                version = int(connection.execute("PRAGMA user_version").fetchone()[0])
                quick_check = str(connection.execute("PRAGMA quick_check").fetchone()[0])
            if version > SCHEMA_VERSION:
                checks.append(
                    Diagnostic(
                        "error",
                        "database schema",
                        f"Schema {version} is newer than supported schema {SCHEMA_VERSION}.",
                    )
                )
            elif version < SCHEMA_VERSION:
                message = (
                    f"Schema {version} will migrate to {SCHEMA_VERSION} "
                    "on the next database command."
                )
                checks.append(Diagnostic("warn", "database schema", message))
            else:
                checks.append(
                    Diagnostic("ok", "database schema", f"Schema {version} is current.")
                )
            level = "ok" if quick_check == "ok" else "error"
            checks.append(Diagnostic(level, "database integrity", quick_check))
        except sqlite3.Error as exc:
            checks.append(Diagnostic("error", "database", f"Cannot inspect {database}: {exc}"))
        permission = _permission_check(database, "database")
        if permission:
            checks.append(permission)
    else:
        checks.append(
            Diagnostic("warn", "database", f"Database does not exist yet: {database}")
        )

    report_directory = expand_path(config["storage"]["report_directory"])
    writable = _writable_parent(report_directory)
    if writable is None or not os.access(writable, os.W_OK):
        checks.append(
            Diagnostic(
                "error",
                "report directory",
                f"No writable parent is available for {report_directory}.",
            )
        )
    else:
        checks.append(
            Diagnostic(
                "ok",
                "report directory",
                f"Reports can be written beneath {report_directory}.",
            )
        )

    if importlib.util.find_spec("jobspy") is None:
        checks.append(
            Diagnostic(
                "error",
                "JobSpy",
                "JobSpy is not importable; reinstall OpenJobScout or run `uv sync`.",
            )
        )
    else:
        checks.append(Diagnostic("ok", "JobSpy", "Discovery dependency is importable."))
    return checks
